Add a custom hook to get calculated data and its status for the Products Reporting page

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

It a pre-tuning PR for #244 and also a follow-up development of #524.
- Add a custom hook to get calculated data and its status for the Products Reporting page

### Detailed test instructions:

💡. For easier mock testing, I pushed another branch `feature/244-products-report-hook-mock-local` with some mock code, which located at:
- Trigger data fetching and print log
    https://github.com/woocommerce/google-listings-and-ads/blob/aaeed073d6ed22c41e992fb1bb2559003a65df8d/js/src/reports/products/index.js#L49-L51
- Mock paid data
    https://github.com/woocommerce/google-listings-and-ads/blob/aaeed073d6ed22c41e992fb1bb2559003a65df8d/src/API/Site/Controllers/Ads/ReportsController.php#L84-L94
- Mock free data
    https://github.com/woocommerce/google-listings-and-ads/blob/aaeed073d6ed22c41e992fb1bb2559003a65df8d/src/API/Site/Controllers/MerchantCenter/ReportsController.php#L83-L90

🩺 . Test with the UIs
1. Checkout `feature/244-products-report-hook-mock-local` branch
2. Head to the Products Reporting page: `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Freports%2Fproducts`
3. Data aren't connected to the UI yet, please open browser DevTool to view the fetched data from the Console panel
4. Made some change on **Date Range** or **Show** UIs, it should reflect with query change and then triggering relevant data fetching, but duplicate query would not re-trigger unnecessary request.
    ![image](https://user-images.githubusercontent.com/17420811/116854209-eb860e80-ac29-11eb-80ce-f60d87f1d3b7.png)

eason9487 committed:
Add a custom hook for getting products reporting data

ecgan reviewed (commented):
I tested based on the test instruction, it seems working okay as per the expectation described. Leaving a comment, and also an observation from my testing which I'm not sure whether it is related or not - might be my local environment bug, or backend bug or mock bug.

eason9487 committed:
Avoid enumerating the dependency query parameters in the hook

To address this comment https://github.com/woocommerce/google-listings-and-ads/pull/540#discussion_r625285098

eason9487 committed:
Let deps of useProductsReport compliant with react hook convention

Addressed the comment https://github.com/woocommerce/google-listings-and-ads/pull/540#discussion_r625948047

ecgan reviewed (approved):
LGTM 🙂  👍
